File: src/04_Arrays_Strings/String_Manipulation.java
/**
 * Level 4: Arrays, Strings & Memory Storage
 * File: String_Manipulation.java
 * Topic: String immutability vs StringBuilder performance for DSA algorithms
 */

public class String_Manipulation {
    public static void main(String[] args) {
        // String Immutability Demo
        String s = "Pre-DSA";
        String s2 = s + " Mastery"; // Creates NEW string object on heap!
        System.out.println("Original String s: " + s);
        System.out.println("New String s2: " + s2);

        // StringBuilder (MUTABLE - Use this in DSA loops!)
        StringBuilder sb = new StringBuilder("Pre-DSA");
        sb.append(" Mastery");
        sb.setCharAt(0, 'p');

        System.out.println("StringBuilder Result: " + sb.toString());
        System.out.println("Reversed StringBuilder: " + sb.reverse().toString());
    }
}

File: src/04_Arrays_Strings/String_Manipulation.cpp
/**
 * Level 4: Arrays, Strings & Memory Storage
 * File: String_Manipulation.cpp
 * Topic: std::string mutability, substrings, and character array mechanics
 */

#include <iostream>
#include <string>

int main() {
    std::string s = "Pre-DSA";
    
    // C++ std::string is MUTABLE in-place:
    s += " Mastery";
    s[0] = 'p'; // Direct element mutation

    std::cout << "Mutated String: " << s << "\n";
    std::cout << "Substring (0, 7): " << s.substr(0, 7) << "\n";
    std::cout << "Length: " << s.length() << "\n";

    return 0;
}
